src/plots/cartesian/index.js

```diff
@@ -189,41 +195,60 @@ function plotOne(gd, plotinfo, cdSubplot, transitionOpts, makeOnCompleteCallback
 
-    for(var i = 0; i < modules.length; i++) {
-        _module = modules[i];
-        var name = _module.name;
-        var categories = Registry.modules[name].categories;
-
-        if(categories.svg) {
-            var className = (_module.layerName || name + 'layer');
-            var plotMethod = _module.plot;
-
-            // plot all visible traces of this type on this subplot at once
-            cdModuleAndOthers = getModuleCalcData(cdSubplot, plotMethod);
-            cdModule = cdModuleAndOthers[0];
-            // don't need to search the found traces again - in fact we need to NOT
-            // so that if two modules share the same plotter we don't double-plot
-            cdSubplot = cdModuleAndOthers[1];
-
-            if(cdModule.length) {
-                layerData.push({
-                    i: traceLayerClasses.indexOf(className),
-                    className: className,
-                    plotMethod: plotMethod,
-                    cdModule: cdModule
-                });
-            }
+    // Plot each zindex group in ascending order
+    var zindices = Object.keys(traceZindexGroups)
+        .map(Number)
+        .sort(function(a, b) { return a - b; });
+    for(var z = 0; z < zindices.length; z++) {
+        var zindex = zindices[z];
+        // For each "module" (trace type)
+        for(var i = 0; i < modules.length; i++) {
+            _module = modules[i];
+            var name = _module.name;
+            var categories = Registry.modules[name].categories;
+
+            if(categories.svg) {
+                var className = (_module.layerName || name + 'layer') + (zindex ? '-' + z : '');
+                var plotMethod = _module.plot;
+
+                // plot all visible traces of this type on this subplot at once
+                cdModuleAndOthers = getModuleCalcData(cdSubplot, plotMethod, zindex);
+                cdModule = cdModuleAndOthers[0];
+                // don't need to search the found traces again - in fact we need to NOT
+                // so that if two modules share the same plotter we don't double-plot
+                cdSubplot = cdModuleAndOthers[1];
+
+                if(cdModule.length) {
+                    layerData.push({
+                        i: traceLayerClasses.indexOf(className),
+                        zindex: z,
+                        className: className,
+                        plotMethod: plotMethod,
+                        cdModule: cdModule
+                    });
+                }
 
-            if(categories.zoomScale) {
-                zoomScaleQueryParts.push('.' + className);
+                if(categories.zoomScale) {
+                    zoomScaleQueryParts.push('.' + className);
+                }
             }
         }
     }
-
-    layerData.sort(function(a, b) { return a.i - b.i; });
+    // Sort the layers primarily by z, then by i
+    layerData.sort(function(a, b) { return a.zindex - b.zindex || a.i - b.i; });
```

Comment:
We should possibly improve condition checking to handle `undefined` cases of `zindex`. 
Right now if `a.zindex = 1` and `b.zindex = undefined` the is `a.zindex - b.zindex` returns `NaN` which is a falsy value. And similarly when `a.zindex = -1` and `b.zindex = undefined` we get the same behavior. Is that correst?
Or should we use something like the following?
```js
    layerData.sort(function(a, b) { return (a.zindex || 0) - (b.zindex || 0) || a.i - b.i; });
```

Reply:
@archmoj `zindex` defaults to 0, so it should never be undefined.

Unless you think it should not have a default, and be undefined if not explicitly set. I like the predictability of defaulting to 0 though.

Reply:
What would happen in the case of traces that do not have `zindex` attribute?
Won't we get here?

Reply:
I've added `zindex` to all cartesian traces, although I might have missed some, so they should have that attribute. Anyways, I added your suggestion just in case, @archmoj

Reply:
It looks like we don't get ``zindex: undefined` here now. But let's keep it safe.
Resolving...